src/address_allocator.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright © 2022 Amazon.com, Inc. or its aFFiliates. All Rights Reserved.
+// Copyright © 2022 Alibaba Cloud. All rights reserved.
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for memory slots.
+//!
+//! This module implements an allocation strategies for memory slots in an
+//! address space (for example MMIO and PIO).
+
+use crate::allocation_engine::{IntervalTree, NodeState};
+use crate::{AllocPolicy, Constraint, Error, Range, Result};
+
+/// Internal representation of AddressAllocator. Contains the managed address
+/// space represented through an instance of Range. The address allocator also
+/// contains a node that represents the root of the interval tree used for
+/// memory slots management. The reason we chose to use an interval tree is
+/// that the average complexity for deletion and insertion is O(log N) and for
+/// searching a node is O(N).
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct AddressAllocator {
+    // Address space that we want to manage.
+    address_space: Range,
+    // Internal representation of the managed address space. Each node in the
+    // tree will represent a memory location and can have two states either
+    // `NodeState::Free` or `NodeState::Allocated`.
+    interval_tree: IntervalTree,
+}
+
+impl AddressAllocator {
+    /// Creates a new instance of AddressAllocator that will be used to manage
+    /// the allocation and release of memory slots from the managed address
+    /// space.
+    #[allow(dead_code)]
+    pub fn new(base: u64, size: u64) -> std::result::Result<Self, Error> {
+        if size == 0 {
+            return Err(Error::InvalidRange(base, base));
+        }
+        let end_option = base.checked_add(size - 1);
+        if let Some(end) = end_option {
+            let aux_range = Range::new(base, end)?;
+            return Ok(AddressAllocator {
+                address_space: aux_range,
+                interval_tree: IntervalTree::new(aux_range, NodeState::Free),
+            });
+        }
+        Err(Error::InvalidRange(base, u64::MAX))
+    }
+
+    /// Allocates a new aligned memory slot.
+    #[allow(dead_code)]
+    pub fn allocate(
+        &mut self,
+        size: u64,
+        align_size: Option<u64>,
```

Comment:
nit: can you specify what is the default and why is this an option? Also, I think we should call this alignment instead of `align_size`.